geom/twkb_parser.go

```diff
@@ -0,0 +1,554 @@
+	}
+	if extprec&2 != 0 {
+		p.hasM = true
+		p.precM = int(DecodeZigZagInt32(uint32(extprec>>5) & 0x07))
+	}
+	if p.hasZ && p.hasM {
+		p.ctype = DimXYZM
+		p.dimensions = 4
+		p.scalings[2] = math.Pow10(-p.precZ)
+		p.scalings[3] = math.Pow10(-p.precM)
+	} else if p.hasZ {
+		p.ctype = DimXYZ
+		p.dimensions = 3
+		p.scalings[2] = math.Pow10(-p.precZ)
+	} else if p.hasM {
+		p.ctype = DimXYM
+		p.dimensions = 3
+		p.scalings[2] = math.Pow10(-p.precM)
+	}
+	return nil
+}
+
+func (p *TWKBParser) parseSize() error {
+	if len(p.twkb) <= p.pos {
+		return p.parserError("cannot unmarshal lacking size varint")
+	}
+	bytesRemaining, err := p.parseUnsignedVarint()
+	if err != nil {
+		return p.parserError("cannot unmarshal size varint malformed")
+	}
+	if uint64(p.pos)+bytesRemaining > uint64(len(p.twkb)) {
+		return p.parserError("cannot unmarshal remaining input smaller than size varint indicates")
+	}
+	return nil
+}
+
+func (p *TWKBParser) parseBBox() error {
+	// Parse the bounding box, but do not otherwise use it.
+	for i := 0; i < p.dimensions; i++ {
+		minVal, err := p.parseSignedVarint()
+		if err != nil {
+			return p.parserError("cannot unmarshal BBox min varint malformed")
+		}
+		p.bbox = append(p.bbox, minVal)
+
+		deltaVal, err := p.parseSignedVarint()
+		if err != nil {
+			return p.parserError("cannot unmarshal BBox delta varint malformed")
+		}
+		p.bbox = append(p.bbox, deltaVal)
+	}
+	return nil
+}
+
+func (p *TWKBParser) parsePoint() (Point, error) {
+	if p.isEmpty {
+		return NewEmptyPoint(p.ctype), nil
+	}
+	return p.nextPoint()
+}
+
+func (p *TWKBParser) nextPoint() (Point, error) {
+	var c Coordinates
+	c.Type = p.ctype
+
+	coords, err := p.parsePointArray(1)
+	if err != nil {
+		return Point{}, err
+	}
+
+	c.XY.X = coords[0]
+	c.XY.Y = coords[1]
+	if p.hasZ && p.hasM {
+		c.Z = coords[2]
+		c.M = coords[3]
+	} else if p.hasZ {
+		c.Z = coords[2]
+	} else if p.hasM {
+		c.M = coords[2]
+	}
+
+	pt, err := NewPoint(c, p.opts...)
+	return pt, p.annotateError(err)
+}
+
+func (p *TWKBParser) parseLineString() (LineString, error) {
+	if p.isEmpty {
+		return NewLineString(NewSequence(nil, p.ctype), p.opts...)
+	}
+	return p.nextLineString()
+}
+
+func (p *TWKBParser) nextLineString() (LineString, error) {
+	coords, err := p.parsePointCountAndArray()
+	if err != nil {
+		return LineString{}, err
+	}
+	ls, err := NewLineString(NewSequence(coords, p.ctype), p.opts...)
+	return ls, p.annotateError(err)
+}
+
+func (p *TWKBParser) parsePolygon() (Polygon, error) {
+	if p.isEmpty {
+		return NewPolygon(nil, p.opts...)
+	}
+	return p.nextPolygon()
+}
+
+func (p *TWKBParser) nextPolygon() (Polygon, error) {
+	numRings, err := p.parseUnsignedVarint()
+	if err != nil {
+		return Polygon{}, p.parserError("cannot unmarshal num rings varint malformed")
+	}
+
+	var rings []LineString
+	for r := uint64(0); r < numRings; r++ {
+		coords, err := p.parsePointCountAndArray()
+		if err != nil {
+			return Polygon{}, err
+		}
+		if len(coords) > 0 {
```

Comment:
This part is of concern to me.
See multiple polygon examples at https://github.com/TWKB/twkb.js/blob/master/test/twkb.spec.js#L43 which seem to do things differently, and particularly https://github.com/TWKB/twkb.js/blob/master/test/twkb.spec.js#L169 but compare that with the spec and you can see why I'm concerned.

Reply:
It seems that at least one implementation emits closed rings when encoding (note there are 5 points given for each ring):
```
=> select ST_AsTWKB('POLYGON((0 0,2 0,2 2,0 2,0 0),(0 0,0 1,1 1,1 0,0 0))'::geometry);
                      st_astwkb                       
------------------------------------------------------
 \x03000205000004000004030000030500000002020000010100
```
When decoding, both closed and unclosed rings are accepted and correctly handled as below:
```
=> select ST_AsText(ST_GeomFromTWKB(E'\\x03000205000004000004030000030500000002020000010100'));
                      st_astext                       
------------------------------------------------------
 POLYGON((0 0,2 0,2 2,0 2,0 0),(0 0,0 1,1 1,1 0,0 0))

=> select ST_AsText(ST_GeomFromTWKB(E'\\x030002040000040000040300040003000202000001'));
                      st_astext                       
------------------------------------------------------
 POLYGON((0 0,2 0,2 2,0 2,0 0),(0 0,0 1,1 1,1 0,0 0))
```
So I think the parser should handle both closed and unclosed rings.

Reply:
I've now implemented that change, to allow the parser to accept either closed or unclosed rings.

Reply:
Update: I'm fairly convinced that the currently implemented approach is correct. That is, if a polygon ring's final point is the same as its first point, the TWKB decoder silently allow that, but also allow the case that the final point isn't the same as its first point (in which case close the ring by appending the first point as its final point). This seems to be what PostGIS does.

My implementation also allows both forms to be generated from the TWKB encoder, with a Boolean flag to control that. Obviously, for size of output it's generally better to omit the final point of a ring as it is redundant information. This seems not to be what PostGIS does, it seems to always output the final point, and there doesn't seem to be a way to control this behaviour. This simply means my implementation is a little more versatile, which is not a problem.